File: src/CCVTAC.Console/History.cs
using System.IO;
using System.Text.Json;
using Spectre.Console;

namespace CCVTAC.Console;

/// <summary>
/// Handles storing, retrieving, and (eventually) analyzing data relating
/// to URLs that the user has entered.
/// </summary>
public class History
{
    private static readonly char Separator = ';';
    private string FilePath { get; init; }
    private byte DisplayCount { get; init; }

    public History(string filePath, byte displayCount)
    {
        FilePath = filePath;
        DisplayCount = displayCount;
    }

    /// <summary>
    /// Add a URL and related data to the history file.
    /// </summary>
    public void Append(string url, DateTime entryTime, Printer printer)
    {
        try
        {
            string serializedEntryTime = JsonSerializer.Serialize(entryTime).Replace("\"", "");
            File.AppendAllText(FilePath, serializedEntryTime + Separator + url + Environment.NewLine);

            printer.Debug($"Added \"{url}\" to the history log.");
        }
        catch (Exception ex)
        {
            printer.Error($"Could not append URL(s) to history log: " + ex.Message);
        }
    }

    public void ShowRecent(Printer printer)
    {
        try
        {
            IEnumerable<IGrouping<DateTime, string>> historyData =
                File.ReadAllLines(FilePath)
                    .TakeLast(DisplayCount)
                    .Select(line => line.Split(Separator))
                    .Where(lineItems => lineItems.Length == 2) // Only lines with date-times
                    .GroupBy(line =>DateTime.Parse(line[0]),
                             line => line[1]);

            Table table = new();
            table.Border(TableBorder.None);
            table.AddColumns("Time", "URL");
            table.Columns[0].PadRight(3);

            foreach (IGrouping<DateTime, string> thisDate in historyData)
            {
                var formattedTime = $"{thisDate.Key:yyyy-MM-dd HH:mm:ss}";
                var urls = string.Join(Environment.NewLine, thisDate);
                table.AddRow(formattedTime, urls);
            }

            printer.PrintTable(table);
        }
        catch (Exception ex)
        {
            printer.Error($"Could not display recent history: {ex.Message}");
        }
    }
}

File: src/CCVTAC.FSharp/History.fs
namespace CCVTAC.Console

open System
open System.IO
open System.Text.Json
open Spectre.Console

type History(filePath: string, displayCount: int) =

    let separator = ';'

    member private _.FilePath = filePath
    member private _.DisplayCount = displayCount

    /// Write a URL and its related data to the history file.
    member this.Append(url: string, entryTime: DateTime, printer: Printer) : unit =
        try
            let serializedEntryTime = JsonSerializer.Serialize(entryTime).Replace("\"", "")
            // TODO: IO should be placed elsewhere.
            File.AppendAllText(this.FilePath, serializedEntryTime + string separator + url + String.newLine)
            printer.Debug $"Added \"%s{url}\" to the history log."
        with exn ->
            printer.Error $"Could not append URL(s) to history log: {exn.Message}"

    member this.ShowRecent(printer: Printer) : unit =
        try
            let lines =
                File.ReadAllLines this.FilePath
                |> Seq.rev
                |> Seq.truncate this.DisplayCount
                |> Seq.rev
                |> Seq.toList

            let historyData =
                lines
                |> Seq.choose (fun line ->
                    match line.Split separator with
                    | [| date; url |] -> Some (DateTime.Parse date, url)
                    | _ -> None)
                |> Seq.groupBy fst
                |> Seq.map (fun (dt, pairs) -> dt, pairs |> Seq.map snd |> Seq.toList)

            // TODO: These presentation matters shouldn't be here.
            let table = Table()
            table.Border <- TableBorder.None
            table.AddColumns("Time", "URL") |> ignore
            table.Columns[0].PadRight(3) |> ignore

            for dateTime, urls in historyData do
                let formattedTime = sprintf "%s" (dateTime.ToString("yyyy-MM-dd HH:mm:ss"))
                let joinedUrls = String.Join(String.newLine, urls)
                table.AddRow(formattedTime, joinedUrls) |> ignore

            Printer.PrintTable table
        with exn ->
            printer.Error $"Could not display history: %s{exn.Message}"
